File: server/internal/service/category.go
package service

import (
	"context"

	"cashcompass-server/internal/model"
	"cashcompass-server/internal/repository"
)

type CategoryService interface {
	List(ctx context.Context) ([]model.Category, error)
	GetByID(ctx context.Context, id int) (model.Category, error)
	Create(ctx context.Context, label string) (model.Category, error)
	UpdateLabel(ctx context.Context, id int, label string) error
	Delete(ctx context.Context, id int) error
}

type CategoryServiceImpl struct {
	categories   repository.CategoryRepository
	transactions repository.TransactionRepository
}

func NewCategoryService(categories repository.CategoryRepository, transactions repository.TransactionRepository) *CategoryServiceImpl {
	return &CategoryServiceImpl{categories: categories, transactions: transactions}
}

func (s *CategoryServiceImpl) List(ctx context.Context) ([]model.Category, error) {
	return s.categories.List(ctx)
}

func (s *CategoryServiceImpl) GetByID(ctx context.Context, id int) (model.Category, error) {
	return s.categories.GetByID(ctx, id)
}

func (s *CategoryServiceImpl) Create(ctx context.Context, label string) (model.Category, error) {
	return s.categories.Create(ctx, label)
}

func (s *CategoryServiceImpl) UpdateLabel(ctx context.Context, id int, label string) error {
	if err := s.categories.UpdateLabel(ctx, id, label); err != nil {
		return err
	}
	if s.transactions != nil {
		return s.transactions.SyncCategoryLabel(ctx, id, label)
	}
	return nil
}

func (s *CategoryServiceImpl) Delete(ctx context.Context, id int) error {
	return s.categories.Delete(ctx, id)
}

File: src/services/category.py
from __future__ import annotations
from src.models import Category
from src.repository.category import CategoryRepository
from src.repository.transaction import TransactionRepository


class CategoryService:
    def __init__(self, repo: CategoryRepository, txn_repo: TransactionRepository):
        self.repo = repo
        self.txn_repo = txn_repo

    @staticmethod
    def _normalize_label(label: str) -> str:
        normalized = label.strip()
        if not normalized:
            raise ValueError("category label is required")
        return normalized

    def list(self) -> list[Category]:
        return self.repo.list()

    def get_by_id(self, id: int) -> Category:
        return self.repo.get_by_id(id)

    def create(self, label: str) -> Category:
        return self.repo.create(self._normalize_label(label))

    def update_label(self, id: int, label: str):
        label = self._normalize_label(label)
        self.repo.update_label(id, label)
        self.txn_repo.sync_category_label(id, label)

    def delete(self, id: int):
        self.txn_repo.clear_category(id)
        self.repo.delete(id)
